ui/src/app/views/workspace-list.tsx

```diff
@@ -83,66 +84,108 @@ const WorkspaceCardMenu: React.FunctionComponent<{
+                    onClick={() => {navigate([wsPathPrefix, 'edit']); }}
                     disabled={!wp.canWrite}>
             Edit
           </MenuItem>
         </TooltipTrigger>
         <TooltipTrigger content={<div>Requires Owner Permission</div>}
                         disabled={wp.isOwner}>
-          <MenuItem onClick={onShare} disabled={!wp.isOwner}>
+          <MenuItem data-test-id='share-workspace' onClick={onShare} disabled={!wp.isOwner}>
             Share
           </MenuItem>
         </TooltipTrigger>
         <TooltipTrigger content={<div>Requires Owner Permission</div>}
                         disabled={wp.isOwner}>
-          <MenuItem onClick={onDelete} disabled={!wp.isOwner}>
+          <MenuItem data-test-id='delete-workspace' onClick={onDelete} disabled={!wp.isOwner}>
             Delete
           </MenuItem>
         </TooltipTrigger>
       </React.Fragment>}
   >
-    <Clickable disabled={disabled} data-test-id='workspace-menu'>
+    <Clickable disabled={disabled} data-test-id='workspace-card-menu'>
       <ClrIcon shape='ellipsis-vertical' size={21}
                style={{color: disabled ? '#9B9B9B' : '#216FB4', marginLeft: -9,
                  cursor: disabled ? 'auto' : 'pointer'}}/>
     </Clickable>
   </PopupTrigger>;
 };
 
-export class WorkspaceCard extends React.Component<
-    {wp: WorkspacePermissions, userEmail: string, reload: Function},
-    {sharing: boolean, confirmDeleting: boolean, workspaceDeletionError: boolean,
-      bugReportOpen: boolean, bugReportError: string}> {
+interface WorkspaceCardState {
+  bugReportError: string;
+  bugReportOpen: boolean;
+  confirmDeleting: boolean;
+  // Whether this card is busy loading data specific to the workspace.
+  loadingData: boolean;
+  sharing: boolean;
+  // The list of user roles associated with this workspace. Lazily populated
+  // only when the workspace share dialog is opened.
+  userRoles?: UserRole[];
+  workspaceDeletionError: boolean;
+}
+
+interface WorkspaceCardProps {
+  userEmail: string;
+  wp: WorkspacePermissions;
+  // A reload function that can be called by this component to reqeust a refresh
+  // of the workspace contained in this card.
+  reload(): Promise<void>;
+}
 
+export class WorkspaceCard extends React.Component<WorkspaceCardProps, WorkspaceCardState> {
   constructor(props) {
     super(props);
     this.state = {
-      bugReportOpen: false,
       bugReportError: '',
-      sharing: false,
+      bugReportOpen: false,
       confirmDeleting: false,
+      loadingData: false,
+      sharing: false,
+      userRoles: null,
       workspaceDeletionError: false
     };
   }
 
   async deleteWorkspace() {
     const {wp} = this.props;
-    workspacesApi().deleteWorkspace(wp.workspace.namespace, wp.workspace.id).then(() => {
-      this.setState({confirmDeleting: false});
-      this.props.reload();
-    }).catch(() => {
-      this.setState({workspaceDeletionError: true});
+    this.setState({
```

Comment:
This was kind of a drive-by fix... deleting a workspace is now a bit cleaner, and we use the same spinner on the workspace card while the deletion API call is pending.